cpp/src/arrow/compute/kernels/scalar_validity.cc

```diff
@@ -76,6 +77,22 @@ struct IsInfOperator {
 
 struct IsNullOperator {
   static Status Call(KernelContext* ctx, const Scalar& in, Scalar* out) {
+    bool is_nan_value = false;
+    bool is_floating = false;
+    if (in.type == float32()) {
+      is_nan_value = std::isnan(internal::UnboxScalar<FloatType>::Unbox(in));
+      is_floating = true;
+    } else if (in.type == float64()) {
+      is_nan_value = std::isnan(internal::UnboxScalar<DoubleType>::Unbox(in));
+      is_floating = true;
+    }
```

Comment:
This can probably made much simpler, e.g.:
```c++
bool* out_value = &checked_cast<BooleanScalar*>(out)->value;
if (in.type->id() == Type::FLOAT) {
  *out_value = !options.nan_is_null || !std::isnan(internal::UnboxScalar<FloatType>::Unbox(in));
}
else if (in.type->id() == Type::DOUBLE) {
  *out_value = !options.nan_is_null || !std::isnan(internal::UnboxScalar<DoubleType>::Unbox(in));
}
else {
  *out_value = true;
}
return Status::OK();
```

Reply:
I would add a check for validity prior to checking float type to prevent unnecessary checks when null bit is enabled. Nit: Also, I would change logic to to use `&&` instead of `||`, feels a bit more readable (and less characters).
```c++
    bool* out_value = &checked_cast<BooleanScalar*>(out)->value;
    if (in.is_valid) {
      switch (in.type->id()) {
        case Type::FLOAT:
          *out_value = options.nan_is_null && std::isnan(internal::UnboxScalar<FloatType>::Unbox(in));
        case Type::DOUBLE:
          *out_value = options.nan_is_null && std::isnan(internal::UnboxScalar<DoubleType>::Unbox(in));
        default:
          *out_value = false;
      }
    } else {
      *out_value = true;
    }
```

Reply:
@pitrou I would be curious if it is worth it to have a specialized version for floating point types (similar to how arithmetic kernels are implemented), as the `options.nan_is_null` check does not applies to other data types.

Reply:
This is the version for Scalar inputs. Micro-optimizing it is futile. The version for Array inputs may be more interesting ;-)